libnavigation-core/src/main/java/com/mapbox/navigation/core/copilot/HistoryAttachmentsUtils.kt

```diff
@@ -0,0 +1,86 @@
+import java.io.File
+import java.text.SimpleDateFormat
+import java.util.Base64.getDecoder
+import java.util.Date
+import java.util.Locale
+import java.util.TimeZone
+
+internal object HistoryAttachmentsUtils {
+
+    private const val HISTORY_FILENAME_SEPARATOR = "__"
+    private const val SDK_PLATFORM = "android"
+    private const val PBF = "pbf"
+    private const val GZ = "gz"
+    private const val PBF_GZ_FORMAT = ".$PBF.$GZ"
+    private const val COPILOT = "co-pilot"
+    private const val PATH_SEPARATOR = "/"
+    private const val HYPHEN = "-"
+    private const val UNDERSCORE = "_"
+
+    fun generateFilename(copilotMetadata: CopilotMetadata): String =
+        "${copilotMetadata.startedAt}$HISTORY_FILENAME_SEPARATOR${copilotMetadata.endedAt}" +
+            "$HISTORY_FILENAME_SEPARATOR$SDK_PLATFORM$HISTORY_FILENAME_SEPARATOR" +
+            "${copilotMetadata.navSdkVersion}$HISTORY_FILENAME_SEPARATOR" +
+            "${copilotMetadata.navNativeSdkVersion}$HISTORY_FILENAME_SEPARATOR" +
+            "$UNDERSCORE$HISTORY_FILENAME_SEPARATOR" +
+            "${copilotMetadata.appVersion}$HISTORY_FILENAME_SEPARATOR" +
+            "${copilotMetadata.appUserId}$HISTORY_FILENAME_SEPARATOR" +
+            "${copilotMetadata.appSessionId}$PBF_GZ_FORMAT"
+
+    fun retrieveOwnerFrom(accessToken: String): String =
+        decode(
+            accessToken.splitToSequence(".").drop(1).first().replace('-', '+').replace('_', '/'),
+        ).getString("u")
+
+    fun generateSessionId(copilotMetadata: CopilotMetadata, owner: String): String =
+        "$COPILOT$PATH_SEPARATOR$owner$PATH_SEPARATOR${retrieveSpecVersion()}$PATH_SEPARATOR" +
+            "${copilotMetadata.appMode}$PATH_SEPARATOR$HYPHEN$PATH_SEPARATOR$HYPHEN" +
+            "$PATH_SEPARATOR${copilotMetadata.driveMode}$PATH_SEPARATOR$HYPHEN" +
+            "$PATH_SEPARATOR${copilotMetadata.driveId}"
+
+    fun retrieveSpecVersion(): String = "1.1"
+
+    fun retrieveNavSdkVersion(): String = BuildConfig.MAPBOX_NAVIGATION_VERSION_NAME
+
+    fun retrieveNavNativeSdkVersion(): String =
+        BuildConfig.NAV_NATIVE_SDK_VERSION
+
+    fun utcTimeNow(format: String, locale: Locale): String {
+        val formatter = SimpleDateFormat(format, locale)
+        formatter.timeZone = TimeZone.getTimeZone("UTC")
+        return formatter.format(Date())
+    }
+
+    fun delete(file: File): Boolean = file.delete()
+
+    fun copyToAndRemove(from: File, filename: String): File {
+        val to = from.copyTo(File(from.parent, filename))
+        from.delete()
+        return to
+    }
+
+    @SuppressLint("NewApi")
+    private fun decode(str: String): JSONObject {
+        val requiredLength = (str.length - 1) / 4 * 4 + 4
+        return if (BuildConfigUtils.retrieveVersionSDK() >= Build.VERSION_CODES.O) {
+            JSONObject(String(getDecoder().decode(str.padEnd(requiredLength, '='))))
+        } else {
+            JSONObject(
+                String(
+                    Base64.decode(
```

Comment:
We can use `android.util.Base64` in all API levels, it is not deprecated.

Reply:
In tests you can do something like this:
```kotlin
mockkStatic(Base64::class)
every { Base64.decode(any<String>(), any()) } answers {
    getDecoder().decode(firstArg<String>())
}
```

Reply:
Let me try that out 👍

Reply:
It works 🎉 

TIL